Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/EpisodeDataManager.swift

```
@@ -1014,3 +1014,84 @@ extension EpisodeDataManager {
     }
 }
+
+// MARK: - Swift Concurrency
+
+extension EpisodeDataManager {
+    @discardableResult
+    func bulkSave(showInfo: [String: String], dbQueue: FMDatabaseQueue) async -> Bool {
+        return await withCheckedContinuation { continuation in
+            dbQueue.inDatabase { db in
+                do {
+                    db.beginTransaction()
+
+                    for episode in showInfo {
+                        try db.executeUpdate("INSERT OR REPLACE INTO EpisodeMetadata VALUES(?, ?);", values: [episode.key, episode.value])
+                    }
+
+                    db.commit()
+                    continuation.resume(returning: true)
+                } catch {
+                    FileLog.shared.addMessage("EpisodeDataManager.bulkSave error: \(error)")
+                    continuation.resume(returning: false)
+                }
+            }
+        }
+    }
+
+    func findEpisodeMetadata(uuid: String, dbQueue: FMDatabaseQueue) async -> Episode.Metadata? {
+        return await withCheckedContinuation { continuation in
+            dbQueue.inDatabase { db in
+                do {
+                    let resultSet = try db.executeQuery("SELECT metadata from EpisodeMetadata WHERE episodeUuid = ?", values: [uuid])
+                    defer { resultSet.close() }
+
+                    if resultSet.next(), let metadataData = resultSet.string(forColumn: "metadata")?.data(using: .utf8) {
+                        Task {
+                            let metadata = await self.getShowInfo(for: metadataData)
+                            continuation.resume(returning: metadata)
+                        }
+                    } else {
+                        continuation.resume(returning: nil)
+                    }
+                } catch {
+                    FileLog.shared.addMessage("EpisodeDataManager.loadMultiple Episode error: \(error)")
+                    continuation.resume(returning: nil)
```

Is hiding the error and returning nil the best approach here? Shouldn't we make the method throw and decide on the calling site? Or is this an approach that we are following on all of this database methods?

---

Good question. This is the first time we're using an `async` approach here.

I changed the code to throw exceptions in ecee87fca0aa4c6c4f11dee3d0e2d1f16c46e5ec and 7efe1ceb82e67f1328e4df6c727684d54231ec67